File: pythonOOP/task_06_abstract_classes.py
"""
TASK 6: Abstract Base Classes (ABC)
====================================
Difficulty: Advanced

Learn about: ABC module, abstractmethod, abstract properties

PROBLEM:
--------
Create a shape calculation system using Abstract Base Classes.

Requirements:
1. Import ABC and abstractmethod from abc module

2. Create an abstract class 'Shape' with:
   - Abstract method: calculate_area()
   - Abstract method: calculate_perimeter()
   - Abstract property: name
   - Concrete method: display() - prints name, area, and perimeter

3. Create class 'Rectangle' inheriting from Shape:
   - Attributes: length, width
   - Implement all abstract methods/properties
   - name property should return "Rectangle"

4. Create class 'Circle' inheriting from Shape:
   - Attribute: radius
   - Implement all abstract methods/properties
   - name property should return "Circle"
   - Use π = 3.14159

5. Create class 'Triangle' inheriting from Shape:
   - Attributes: side_a, side_b, side_c
   - Implement all abstract methods/properties
   - name property should return "Triangle"
   - Use Heron's formula for area: √[s(s-a)(s-b)(s-c)] where s = (a+b+c)/2

6. Create a function 'compare_shapes(shape1, shape2)' that:
   - Compares two shapes by area
   - Prints which shape is larger or if they're equal

TEST YOUR CODE:
---------------
- Create objects of each shape type
- Calculate and display their areas and perimeters
- Compare different shapes
- Try to instantiate Shape directly (should fail)
"""

# Write your solution below:

File: javaOOP/Task06AbstractClasses.java
/**
 * TASK 6: Abstract Classes
 * =========================
 * Difficulty: Advanced ⭐⭐⭐⭐
 * 
 * Learn about: Abstract classes, Abstract methods, Template method pattern
 * 
 * PROBLEM:
 * --------
 * Create a shape calculation system using abstract classes.
 * 
 * Requirements:
 * 
 * 1. Create an abstract class 'Shape':
 *    - Instance variable: color (String)
 *    - Constructor taking color
 *    - Abstract method: calculateArea() - returns double
 *    - Abstract method: calculatePerimeter() - returns double
 *    - Concrete method: displayInfo() - prints shape info and calls abstract methods
 *    - Concrete method: getColor() - returns color
 * 
 * 2. Create concrete class 'Circle' extends Shape:
 *    - Instance variable: radius (double)
 *    - Constructor taking color and radius
 *    - Implement calculateArea() - π * r²
 *    - Implement calculatePerimeter() - 2 * π * r
 * 
 * 3. Create concrete class 'Rectangle' extends Shape:
 *    - Instance variables: length (double), width (double)
 *    - Constructor taking color, length, width
 *    - Implement calculateArea() - length * width
 *    - Implement calculatePerimeter() - 2 * (length + width)
 * 
 * 4. Create concrete class 'Triangle' extends Shape:
 *    - Instance variables: side1, side2, side3 (double)
 *    - Constructor taking color and three sides
 *    - Implement calculateArea() - use Heron's formula
 *    - Implement calculatePerimeter() - sum of all sides
 *    - Add validation: sides must form valid triangle
 * 
 * TEST YOUR CODE:
 * ---------------
 * - Create different shapes
 * - Store in Shape array (polymorphism)
 * - Calculate total area of all shapes
 * - Display info for each shape
 * 
 * CONCEPTS TO UNDERSTAND:
 * -----------------------
 * - Abstract classes (cannot be instantiated)
 * - Abstract methods (must be implemented by subclasses)
 * - Concrete methods in abstract classes
 * - Template method pattern
 * - When to use abstract classes vs interfaces
 * 
 * EXPECTED OUTPUT EXAMPLE:
 * ------------------------
 * Shape: Circle
 * Color: Red
 * Area: 78.54
 * Perimeter: 31.42
 * 
 * Shape: Rectangle
 * Color: Blue
 * Area: 50.00
 * Perimeter: 30.00
 * 
 * Total area of all shapes: 128.54
 */

// Write your solution below:

